### PR TITLE
add slack reaction based moderation alerts

### LINKED ISSUE #4577: NestBot Content Reporting & Moderation Workflow
Community & Support: [LinkedIn Group](https://www.linkedin.com/groups/14656108/) · [Slack #project-nest](https://owasp.slack.com/archives/project-nest)

Active project leaders: Arkadii Yakovets -- [GitHub](https://github.com/arkid15r/) · [LinkedIn](https://www.linkedin.com/in/arkid15r/) · [Slack](https://owasp.slack.com/team/U060W3NKLTF); Kate Golovanova -- [GitHub](https://github.com/kasya/) · [LinkedIn](https://www.linkedin.com/in/kate-golovanova/) · [Slack](https://owasp.slack.com/team/U07PWB1JZ6Z)

[Contributing](https://github.com/owasp/nest/blob/main/CONTRIBUTING.md) · [Code of Conduct](https://github.com/owasp/nest/blob/main/CODE_OF_CONDUCT.md) · [GSoC Mentors](https://github.com/owasp/nest/blob/main/MENTORS.md)

---

**Is your feature request related to a problem? Please describe.**

We are seeing periodic questionable content (e.g., spam, scams, irrelevant posts) in Slack channels such as #owasp-community, #jobs, and potentially others. This issue proposes implementing a configurable content reporting and moderation workflow via NestBot to enable community-driven flagging and faster response from moderators.

**Describe the solution you'd like**

Goals

* Provide a simple, low-friction mechanism for users to report questionable content
* Enable configurable reporting rules per channel
* Notify appropriate moderators/admins based on severity and thresholds
* Stretch goal: Lay groundwork for optional automated mitigation actions by NestBot

### 1. Emoji-Based Reporting (Primary Mechanism)

Use Slack emoji reactions as the primary reporting signal.

**Key ideas:**

* Specific emojis map to report types:

  * `:spam:` → spam
  * `:scam:` → scam
  * `:abuse:` → inappropriate content

* Configurable per channel:

  * Allowed/reporting emojis
  * Minimum number of unique users reacting before triggering escalation
  * Optional weighting (future enhancement)

**Why emoji-first:**

* Native to Slack, no extra UI friction
* Fast and intuitive
* Socially accepted and lightweight

### 2. Configurable Rules (Per Channel)

Define a flexible configuration model with:

* **Channel scope**
* **Report types (emoji → category mapping)**
* **Thresholds**

  * Minimum number of reports
  * Previous reports for this user content

* **Notification targets**

  * Specific users (admins/moderators)
  * Dedicated moderation channel

* **Cooldowns / deduplication**

  * Prevent repeated alerts for the same message

### 3. Notification Workflow

When a message crosses a configured threshold:

* NestBot sends a structured alert including:

  * Message link
  * Channel
  * Author
  * Report type(s)
  * Number of reports
 
* Notify:
  * Designated moderators/admins
  * Optional logging channel for audit trail

### 4. Optional: Message Shortcuts / “Connect to Apps” Flow

Evaluate adding Slack Message Shortcuts as an alternative reporting mechanism:

**Pros:**

* Explicit reporting flow (form-based input, optional comments)
* More structured metadata

**Cons:**

* Higher friction vs emoji reactions
* Less likely to be used consistently

**Recommendation:**

* Start with emoji-based reporting
* Reassess need for shortcuts after initial rollout and usage data

### 5. Initial Mitigation Actions (Phase 1)

* Notify moderators/admins only
* No automatic enforcement initially

### 6. Future Enhancements (Phase 2+)

Potential automated actions (with strict safeguards and testing):

* Auto-reply warning to message author
* Temporary message hiding (if Slack permissions allow)
* Auto-thread escalation
* Rate limiting repeat offenders
* Integration with user reputation/history

**Important:**
Automated actions must require:

* Careful threshold tuning
* Explicit authorization scopes
* Extensive testing to avoid false positives

## Open Questions

* What default thresholds should we start with?
* Do we require reports from unique users only?
* Should certain roles (e.g., moderators) have higher weight?
* Do we want a centralized moderation dashboard later?
* Are Slack permissions sufficient for any automated actions we may want?

## Acceptance Criteria

* Emoji-based reporting triggers are detected reliably
* Per-channel configuration is supported
* Notifications are sent when thresholds are met
* Duplicate alerts are avoided
* System is extensible for future automation

## Notes

* Prioritize simplicity and adoption over feature richness
* Emoji-based reporting should remain the primary UX unless proven insufficient
* “Golden rule”: validate assumptions with OWASP staff/moderators before rollout

**Are you going to work on implementing this?**

- [ ] Yes
- [x] No

**Additional context**

After PoC is ready we could introduce all messages within a monitored channel to be scanned by NestBot and score for potential content reporting (before users' reports via emoji or other way).

### PR DESCRIPTION
**Pull request by Mr-Rahul-Paul**
## Proposed Change

Resolves #4577.

This PR adds a **Phase 1** implementation for Slack reaction-based content reporting in NestBot.

When a configured reaction reaches its threshold on a Slack message, NestBot records the alert and posts one notification to a configured moderation channel. This phase is notification-only and does not delete, hide, warn, or DM users.

NestBot moderation reaction flow

## What Changed

- Added `ModerationRule` for admin-managed reaction rules and thresholds.
- Added `ModerationAlert` to record sent alerts and prevent duplicates.
- Added a `reaction_added` Slack event handler.
- Added required Slack manifest entries:
  - `reactions:read`
  - `reaction_added`
- Added local ngrok origin support for Slack callback testing.

## Setup Notes

  For local testing, use a development Slack workspace and a local ngrok tunnel.

- Slack needs a public HTTPS callback URL, so `ngrok is required` for local testing. Start ngrok, copy
the `Forwarding` URL, and use that URL in the Slack app manifest for the Slack callback URLs.

- Moderation rules are configured in Django admin.
    - replace the links of `owasp.org...` to the ngrok link in `https://app.slack.com/app-settings/.../.../app-manifest`
    - get the slack tokens from `https://api.slack.com/apps/.../install-on-team` 
    - paste the `Forwarding` ngrok link in `backend/.env`

Required local values include:

```env
DJANGO_SLACK_BOT_TOKEN=...
DJANGO_SLACK_SIGNING_SECRET=...
SLACK_BOT_TOKEN_<WORKSPACE_ID>=`SAME TOKEN AS DJANGO_SLACK_BOT_TOKEN`
DJANGO_LOCAL_NGROK_ORIGIN=https://<ngrok-domain>
````
`SLACK_BOT_TOKEN_<WORKSPACE_ID>` (which is same as `DJANGO_SLACK_BOT_TOKEN`) is needed when running the existing `slack_sync_data` command to create Conversation rows. For a single development workspace, it can reuse the same token value as `DJANGO_SLACK_BOT_TOKEN`.

After updating the Slack manifest, the app must be reinstalled so the new event and scope are active.

## Notes

- Slack emoji aliases are not normalized yet.
- NestBot must be added to both the source channel and the moderation alert channel.
- Local ngrok URLs should not be committed to the manifest.

## todo
~Add unit tests for reaction handling and duplicate-alert behavior.~ **done.**
~Improve validation for moderation rule fields.~ **done.**
~move the logic in a seperate service~ **done.**

## Checklist

* [x] Required: I followed the contributing workflow: https://github.com/OWASP/Nest/blob/main/CONTRIBUTING.md#contributing-workflow
* [x] Required: I verified that my code works as intended and resolves the issue as described.
* [x] Required: I ran `make check-test` locally: all warnings addressed, tests passed.
* [x] I used AI for code, documentation, tests, or communication related to this PR.

### CODE REVIEW

**Comment by Mr-Rahul-Paul** [comment]
i have split the original `handle_event` flow into small helper methods. To handle the api failures (as pointed by code rabbit) The behavior is the same, but the alert row is now created only after the notification succeeds.

**Comment by Mr-Rahul-Paul** [comment]
@arkid15r 
 The existing `slack_sync_data` flow already reads `Workspace.bot_token`, which resolves to `SLACK_BOT_TOKEN_{workspace_id.upper()}`.

That env key was not documented in `.env.example`, so local Slack sync setup was incomplete. I added a dotenv-safe example key so contributors know what to set when they need to run Slack data sync locally.

let me know if you have any issues

**Comment by Mr-Rahul-Paul** [comment]
The Slack sync commands were not reading `DJANGO_SLACK_BOT_TOKEN`. They were implemented to read a dynamic workspace-specific env var: `SLACK_BOT_TOKEN_<WORKSPACE_ID>`

for OWASP it expects something like SLACK_BOT_TOKEN_T04T40NHX. Since that env var was not present, the sync target would skip the workspace and never actually sync to Slack.

 I added the dynamic token to the example env using a placeholder `SLACK_BOT_TOKEN_T1234567890` with a helper comment  , also we need a proxy link as well , `ngrok` to setup the nest bot locally , also added a comment there.
 
The value `SLACK_BOT_TOKEN_<WORKSPACE_ID>` is the same as `DJANGO_SLACK_BOT_TOKEN` (when setup locally atleast). The issue is the env var name. The current implementation seems designed for multi-workspace support, but since Nestbot only needs to manage the OWASP Nest's workspace, I think it would be simpler and less error-prone for the sync commands to read DJANGO_SLACK_BOT_TOKEN directly, or at least fall back to it.

File: backend/apps/slack/management/commands/slack_sync_moderation_rules.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+from pathlib import Path
+
+import yaml
+from django.conf import settings
+from django.core.management.base import BaseCommand, CommandError
+
+from apps.slack.models import Conversation
+from apps.slack.models.moderation import ModerationRule
+
+DEFAULT_CONFIG_PATH = Path(settings.BASE_DIR) / "apps/slack/config/moderation_rules.yaml"
```
Comment (arkid15r) [inline backend/apps/slack/management/commands/slack_sync_moderation_rules.py]: I'm not sure if we need the YAML representation. Would DB record be just enough? What's your reasoning behind adding it?

File: backend/apps/slack/models/moderation.py
Comment (arkid15r) [inline backend/apps/slack/models/moderation.py:1]: It belongs with Slack, no doubt. Let's consider more general naming for this. Yes, we want to solve the spam/moderation problem atm however emoji/reaction mechanism is wider and could be used in some other way -- basically more flexible than just moderation.

File: backend/apps/slack/models/moderation.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    )
+    alert_user_ids = models.JSONField(blank=True, default=list)
+    is_enabled = models.BooleanField(default=True)
+
+    @override
+    def __str__(self):
+        return f"{self.conversation} :{self.emoji_name}"
+
+
+class ModerationAlert(TimestampedModel):
```
Comment (arkid15r) [inline backend/apps/slack/models/moderation.py]: A single class per model file please.

File: backend/apps/slack/services/moderation.py [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+    text = _build_alert_text(rule, channel_id, emoji_name, reaction_count, permalink)
+    alert = _post_alert(client, rule.alert_channel_id, text)
+    if alert is None:
+        moderation_alert.delete()
+        return
+
+    _update_alert_message_ts(moderation_alert, alert)
+
+
+def _get_message_reaction_details(event):
```
Comment (arkid15r) [inline backend/apps/slack/services/moderation.py]: Please don't use `_` prefix.

File: backend/settings/local.py [hunk 4]
```diff
@@ -1,18 +1,27 @@
...
 
 class Local(Base):
     """Local configuration."""
 
     APP_NAME = "OWASP Nest Local"
 
-    ALLOWED_ORIGINS = (
-        "http://127.0.0.1:3000",
-        "http://localhost:3000",
+    ALLOWED_ORIGINS = tuple(
```
Comment (arkid15r) [inline backend/settings/local.py]: Use list and append `LOCAL_NGROK_ORIGIN` value if needed (also consider `:=`).

Comment (arkid15r) [reply backend/settings/local.py]: Why is it needed btw?

File: docker-compose/local/compose.yaml [hunk 5]
```diff
@@ -4,6 +4,7 @@ services:
     command: >
       sh -c '
         python manage.py migrate &&
+        python manage.py slack_sync_moderation_rules &&
```
Comment (arkid15r) [inline docker-compose/local/compose.yaml]: Not sure it's needed. Having those just in the DB is more simple.

**Review by arkid15r — CHANGES_REQUESTED** [review]

File: backend/apps/slack/management/commands/slack_sync_moderation_rules.py (hunk 1 above)
Comment (Mr-Rahul-Paul) [reply backend/apps/slack/management/commands/slack_sync_moderation_rules.py]: I added YAML to provide defaults that would be synchronized into the database, while still allowing admin edits.
but i agree , this creates two sources of truth of the configuration . 
I'll remove the YAML sync command and startup hook.

File: backend/apps/slack/models/moderation.py
Comment (Mr-Rahul-Paul) [reply backend/apps/slack/models/moderation.py:1]: alright I'll rename it to `ReactionRule` and `ReactionAlert`

File: backend/apps/slack/models/moderation.py (hunk 2 above)
Comment (Mr-Rahul-Paul) [reply backend/apps/slack/models/moderation.py]: Will do.

File: backend/apps/slack/management/commands/slack_sync_moderation_rules.py (hunk 1 above)
Comment (Mr-Rahul-Paul) [reply backend/apps/slack/management/commands/slack_sync_moderation_rules.py]: removed.

File: docker-compose/local/compose.yaml (hunk 5 above)
Comment (Mr-Rahul-Paul) [reply docker-compose/local/compose.yaml]: removed.

File: backend/settings/local.py (hunk 4 above)
Comment (Mr-Rahul-Paul) [reply backend/settings/local.py]: I added this while debugging the local Slack/ngrok setup.  I’ve removed it.

File: infrastructure/modules/ecr-cache/.terraform.lock.hcl
Comment (rudransh-shrivastava) [inline infrastructure/modules/ecr-cache/.terraform.lock.hcl:1]: is this change intended?

Comment (Mr-Rahul-Paul) [reply infrastructure/modules/ecr-cache/.terraform.lock.hcl:1]: NO it was not intended , running terraform checks sometimes generates new lock files on my machine 
REMOVED.

Thanks for pointing it out.

File: backend/src/apps/slack/models/reaction_alert.py [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+class ReactionAlert(TimestampedModel):
+    """Record that an alert was already sent for a reported message."""
+
+    class Meta:
+        """Model options."""
+
+        db_table = "slack_reaction_alerts"
+        unique_together = ("conversation", "message_ts", "report_type")
+
+    conversation = models.ForeignKey(Conversation, on_delete=models.CASCADE)
```
Comment (arkid15r) [inline backend/src/apps/slack/models/reaction_alert.py]: Your code is inconsistent with the rest of the code base models.

**Review by arkid15r — APPROVED** [review]

**Review by kasya — APPROVED** [review]
